src/controllers/controller.h

```diff
@@ -169,12 +176,32 @@ class Controller : public QObject, ConstControllerPresetVisitor {
     bool m_bLearning;
     QElapsedTimer m_userActivityInhibitTimer;
 
-    UserSettingsPointer m_pConfig;
-
+    friend class ControllerJSProxy;
     // accesses lots of our stuff, but in the same thread
     friend class ControllerManager;
     // For testing
     friend class ControllerPresetValidationTest;
 };
 
+// An object of this class gets exposed to the JS engine, so the methods of this class
+// constitute the api that is provided to scripts under "controller" object.
+// See comments on ControllerEngineJSProxy.
+class ControllerJSProxy : public QObject {
+    Q_OBJECT
+  public:
+    ControllerJSProxy(Controller* m_pController)
+            : m_pController(m_pController) {
+    }
+
+    // The length parameter is here for backwards compatibility for when scripts
+    // were required to specify it.
+    Q_INVOKABLE virtual void send(QList<int> data, unsigned int length = 0) {
+        Q_UNUSED(length);
```

Comment:
Verify that this functions is not invoked with inconsistent arguments:
DEBUG_ASSERT(length == 0 || data.length() == length)

Not needed, but better verify our assumptions

Reply:
Eh, I'd rather leave it as it is. Nothing will go wrong if the assumption is violated. It's valid for this to be called with the default 0 for the second argument.